### PR TITLE
fix: terminate workflow should not get throttled Fixes #12778

### LINKED ISSUE #12778: Cronworkflow with concurrencyPolicy: Replace/Forbid should not queue up pending workflows
### Pre-requisites

- [X] I have double-checked my configuration
- [X] I can confirm the issue exists when I tested with `:latest`
- [X] I have searched existing issues and could not find a match for this bug
- [x] I'd like to contribute the fix myself (see [contributing guide](https://github.com/argoproj/argo-workflows/blob/main/docs/CONTRIBUTING.md))

### What happened/what did you expect to happen?

When argo-workflow-controller hits parallelism limit,
the new scheduled cronworkflow will all be in pending regardless of the `concurrencyPolicy`

[Concurrent policy](https://argo-workflows.readthedocs.io/en/latest/cron-workflows/#cronworkflow-options)
Policy that determines what to do if multiple Workflows are scheduled at the same time. Available options: Allow: allow all, Replace: remove all old before scheduling a new, Forbid: do not allow any new while there are old

image

Although when the parallelism slot is available, the controller will clear up the pending cronworkflow according to `concurrencyPolicy`
There still exists a problem if we have alerts that depend on the number of pending workflows, we could end up with hundreds of pending workflow that triggers the alert.
I think in an ideal scenario, when a cronworkflow is scheduled, It should check existing scheduled pending workflow as well and operate accordingly.

### Version

v3.5.5

### Paste a small workflow that reproduces the issue. We must be able to run the workflow; don't enter a workflows that uses private images.

```YAML
metadata:
  name: omniscient-rhino
  namespace: argo
spec:
  schedule: '* * * * *'
  concurrencyPolicy: "Replace"
  workflowSpec:
    entrypoint: argosay
    templates:
        container:
          name: main
          image: argoproj/argosay:v2
          command:
            - /argosay
          args:
            - echo
            - hello
```

### Logs from the workflow controller

```text
kubectl logs -n argo deploy/workflow-controller | grep ${workflow}
```

### Logs from in your workflow's wait container

```text
kubectl logs -n argo -c wait -l workflows.argoproj.io/workflow=${workflow},workflow.argoproj.io/phase!=Succeeded
```

### PR DESCRIPTION
**Pull request by tczhao**
Fixes #12778 

### Motivation

To prevent cronworkflow from queueing up pending workflows when
- reached parallelism limit 
- concurrenpolicy-replace already update the spec of the pending workflow to `shutdown: terminated`

### Modifications

Allow workflow in the queue to be processed (operated) regardless of parallelism limit.

We have scenarios
- reach parallelism limit, terminate running workflow (no issue)
- reach parallelism limit, terminate pending workflow (we should not block the queue from processing these workflow, just like semaphore doesn't block the queue from processing pending workflow that got terminated https://github.com/argoproj/argo-workflows/blob/5b3909b9567a5c6fc4fef81ac0785cfddbffb291/workflow/controller/operator.go#L528)

### Verification

Add test,
test fail
Add code change
test succeed

### CODE REVIEW

**Comment by agilgur5** [comment]
> * concurrenpolicy-replace already update the spec of the pending workflow to `shutdown: terminated`

oh interesting, I didn't know that. that does make the fix much simpler.

> * reach parallelism limit, terminate running workflow (no issue)
> * reach parallelism limit, terminate pending workflow (we should not block the queue from processing these workflow, just like semaphore doesn't block the queue from processing pending workflow that got terminated

this makes sense to me and sounds consistent, so I'll approve, but I'm not an expert in the synchronization area of the codebase so would appreciate if someone else could take a look. in particular, technically speaking this change means that terminating Workflows are no longer counted toward parallelism limits, right?

**Review by agilgur5 — APPROVED** [review]

**Comment by agilgur5** [comment]
Also if you're interested, another synchronization race condition issue is #11984

**Comment by tczhao** [comment]
> technically speaking this change means that terminating Workflows are no longer counted toward parallelism limits, right?

Correct,
We either
- terminate running workflow (in parallelism limit, release it from parallelism)
- terminate pending/unknown workflow (bypass parallelism)
- terminate finished workflow (no effect)

File: workflow/controller/controller.go [hunk 1]
```diff
@@ -834,7 +834,7 @@ func (wfc *WorkflowController) processNextItem(ctx context.Context) bool {
 
 	woc := newWorkflowOperationCtx(wf, wfc)
 
-	if !wfc.throttler.Admit(key.(string)) {
+	if !(woc.GetShutdownStrategy().Enabled() && woc.GetShutdownStrategy() == wfv1.ShutdownStrategyTerminate) && !wfc.throttler.Admit(key.(string)) {
```
Comment (sarabala1979) [inline workflow/controller/controller.go:837]: @tczhao Quick suggestion:
Can we handle this state workflow in the informer filter section so, the workflow will not even get into the throttler queue?
https://github.com/argoproj/argo-workflows/blob/59d7e5b129d736ae1f34c51eaccb96dd3cb0d7be/workflow/controller/controller.go#L992

Comment (tczhao) [reply workflow/controller/controller.go:837]: The informer filter only exclude the workflow.
We want to include the shutdown workflow in the wfqueue, and process it.
Here we skip throttle/parallelism check if the workflow is marked shutdown

Comment (sarabala1979) [reply workflow/controller/controller.go:837]: oh yeah got it thanks.  I thought opposite way. make sense

**Review by sarabala1979 — APPROVED** [review]
LGTM

**Comment by agilgur5** [comment]
Backported cleanly to `release-3.5` as c9dd50d35

**Pull request merged**